### PR TITLE
Skip mode validation for snapshots

### LINKED ISSUE #12010: Scylla should skip mode validation of snapshot files
Since 65dd45d9cf83e562993e88dcbf1e214117d46b24 we're verifying the owner and mode of all files in the table directory when scylla is started.
But we do that recursively, diving into the snapshots sub directory.
This is not necessarily a bad thing, however scylla manager might delete snapshot in parallel, causing scylla to fail to start with e.g.:
```
2022-10-01T12:52:49+00:00 longevity-mv-si-4d-2022-2-db-node-4a622274-13      !ERR | scylla[115686]:  [shard  0] database - Exception while populating keyspace 'system_schema' with column family 'view_virtual_columns' from file '/var/lib/scylla/data/system_schema/view_virtual_columns-08843b6345dc3be29798a0418295cfaa': std::filesystem::__cxx11::filesystem_error (error system:2, filesystem error: stat failed: No such file or directory [/var/lib/scylla/data/system_schema/view_virtual_columns-08843b6345dc3be29798a0418295cfaa/snapshots/sm_20220930165632UTC/me-982982-big-Summary.db])
```

See https://github.com/scylladb/scylla-cluster-tests/issues/5431

### PR DESCRIPTION
**Pull request by bhalevy**
Skip over verification of owner and mode of the snapshots
sub-directory as this might race with scylla-manager
trying to delete old snapshots concurrently.

Fixes #12010

### CODE REVIEW

File: replica/distributed_loader.cc [hunk 1]
```diff
@@ -85,8 +85,18 @@ io_error_handler error_handler_gen_for_upload_dir(disk_error_signal_type& dummy)
...
-        return utils::directories::verify_owner_and_mode(d.sstable_dir());
+    // verify owner and mode on the sstables directory
+    // and all its subdirectories, except for "snapshots"
+    // as there could be a race with scylla-manager that might
+    // delete snapshots concurrently
+    co_await dir.invoke_on(0, [] (const sstables::sstable_directory& d) -> future<> {
+        fs::path sstable_dir = d.sstable_dir();
+        co_await utils::directories::verify_owner_and_mode(sstable_dir, utils::directories::recursive::no);
+        co_await lister::scan_dir(sstable_dir, {}, [] (fs::path dir, directory_entry de) -> future<> {
+            if (de.type && *de.type == directory_entry_type::directory && de.name != sstables::snapshots_dir) {
```
Comment (xemul) [inline replica/distributed_loader.cc:96]: The call to `scan_dir` accepts entry type to filter directories, it's cleaner to use it

Comment (bhalevy) [reply replica/distributed_loader.cc:96]: you're right.  I forgot about it

**Comment by bhalevy** [comment]
In v2:
- use `lister::dir_entry_types::of<directory_entry_type::directory>()` for scan_dir for sub-directories

File: utils/directories.cc [hunk 2]
```diff
@@ -128,6 +128,9 @@ future<> directories::verify_owner_and_mode(fs::path path) {
         if (!can_access) {
             verification_error(std::move(path), "Directory cannot be accessed for read, write, and execute");
         }
+        if (level && !recurse) {
+            co_return;
+        }
         co_await lister::scan_dir(path, {}, [] (fs::path dir, directory_entry de) -> future<> {
             co_await verify_owner_and_mode(dir / de.name);
```
Comment (xemul) [inline utils/directories.cc:135]: Shouldn't it call `do_...` version with `level++`ed?

Comment (xemul) [reply utils/directories.cc:135]: Or wait, we don't need `level` at all -- just don't dive into recursion if `!recurse`

Comment (xemul) [reply utils/directories.cc:135]: Ah, it's called for plain files too, so probably we do need level. In this case it should call the `do_...` with increased level as I suggested above

Comment (bhalevy) [reply utils/directories.cc:135]: hmm, that's odd.  I had it this way locally and I probably didn't amend the patch before pushing....

**Comment by bhalevy** [comment]
In v3 ([845b6f9](https://github.com/scylladb/scylladb/commit/845b6f901b071fa1bc92bc01d1e73d19faaaf630)):
- pass `recursive` flag and `level + 1` do nested `do_verify_owner_and_mode` call.

**Comment by bhalevy** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2781/

https://github.com/scylladb/scylladb/issues/14842.
Respinned

**Pull request merged**